src/http.cc

```diff
@@ -826,11 +839,113 @@ HttpStateData::handle1xx(HttpReply *reply)
     // for similar reasons without a 1xx response.
 }
 
+/// Matching protocols predicate function for std::find_if algorithm
+/// It handle the case the reference protocol includes a protocol version
+/// but the checking protocol does not (eg when squid requests upgrade to
+/// 'websockets' and server reports upgrade to 'websocket/1.2').
+class MatchProtocol
+{
+public:
+    explicit MatchProtocol(const char *s, size_t len) :ref(s), refLen(len) {}
+    bool operator() (const SBuf &checking) {
+        const char *p;
+        if (checking.find('/') == SBuf::npos && (p = std::find(ref, ref + refLen, '/')) != ref + refLen)
+            return checking.caseCmp(ref, p - ref) == 0;
+        else
+            return checking.caseCmp(ref, refLen) == 0;
+    }
+
+private:
+    const char *ref;
+    size_t refLen;
+};
+
+bool
+HttpStateData::upgradeProtocolsSupported(String &upgradeProtos)
+{
+    if (!upgradeProtocols)
+        return false;
+
+    const char *pos = NULL;
+    const char *item;
+    int ilen = 0;
+    while (strListGetItem(&upgradeProtos, ',', &item, &ilen, &pos)) {
```

Comment:
I did not convert it to SBuf. 
Now String it is used only inside this method I believe it is better and a little faster to avoid conversion.

The code which extracts headers list items exist in some places inside squid as it is here.

In the other hand using SBuf may be a small step to completely drop String from Squid. But it can be done when SBuf supported by HttpMsg objects.

If you believe it is important I can try use SBuf.